### PR TITLE
Split etcdctl into etcdctl (public API access) & etcdutl (direct surgery on files)

### PR DESCRIPTION
**Pull request by ptabor**
Split etcdctl into etcdctl (public API access) & etcdutl (direct surgery on files)

Motivation is as follows: 
  - etcdctl we only depend on clientv3 APIs, no dependencies of bolt, backend, mvcc, file-layout
  - etcdctl can be officially supported across wide range of versions, while etcdutl is pretty specific to file format at particular version. 
  -  it's step towards desired modules layout, documented in: https://etcd.io/docs/next/dev-internal/modules/

This PR defines new binary: etcdutl with following jobs: 
  -  `(V2) etcdctl backup` -> `etcdutl backup`
  -  `(V3) etcdctl snapshot restore` -> `etcdutl snapshot restore`
  -  `(V3) etcdctl snapshot status` -> `etcdutl snapshot status`
  -  `(V3) etcdctl defrag --date-dir=/..' -> `etcdutl defrag --date-dir=/..`
  -  `(V3) etcdctl migrate` -> I decided to decommission that tool as we support migration by 1 minor version... so continous support of V2->V3 sounds too generous (and the code is very subtle), 

The etcdctl commands keep being served as they used to, but just print additional warning with recommendation to migrate to etcdutl. 

~This PR is on top of: https://github.com/etcd-io/etcd/pull/12969~ [rebased]
It fullfils one one of the goals from: https://github.com/etcd-io/etcd/issues/12330#issuecomment-779091692

### CODE REVIEW

**Comment by gyuho** [comment]
In general, I like this approach (decoupling server-side dependencies for client-side operation). Can we make this more incremental? I think cobrautil can be a separate PR?

**Comment by ptabor** [comment]
Thank you for merging in the base. I rebased the PR.

**Comment by gyuho** [comment]
I believe if we make this transition, this should be landed in 3.5. I am in favor of merging this, as long as @hexfusion @lilic are comfortable, since they are initial 3.5 release managers.

File: etcdutl/README.md [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+
+### JSON
+
+The JSON encoding of the command's [RPC response][etcdrpc]. Since etcd's RPCs use byte strings, the JSON output will encode keys and values in base64.
+
+Some commands without an RPC also support JSON; see the command's `Output` description.
+
+### Protobuf
+
+The protobuf encoding of the command's [RPC response][etcdrpc]. If an RPC is streaming, the stream messages will be concetenated. If an RPC is not given for a command, the protobuf output is not defined.
```
Comment (lilic) [inline etcdutl/README.md:172]:
```suggestion
The protobuf encoding of the command's [RPC response][etcdrpc]. If an RPC is streaming, the stream messages will be concatenated. If an RPC is not given for a command, the protobuf output is not defined.
```

File: etcdutl/README.md [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+```bash
+./etcdutl version
+# etcdutl version: 3.5.0
+# API version: 3.1
+```
+
+
+## Exit codes
+
+For all commands, a successful execution return a zero exit code. All failures will return non-zero exit codes.
```
Comment (lilic) [inline etcdutl/README.md:152]:
```suggestion
For all commands, a successful execution returns a zero exit code. All failures will return non-zero exit codes.
```

File: etcdutl/README.md [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+``` bash
+# Defragment while etcd is not running
+./etcdctl defrag --data-dir default.etcd
+# success (exit status 0)
+# Error: cannot open database at default.etcd/member/snap/db
+```
+
+#### Remarks
+
+DEFRAG returns a zero exit code only if it succeeded defragmenting all given endpoints.
```
Comment (lilic) [inline etcdutl/README.md:37]:
```suggestion
DEFRAG returns a zero exit code only if it succeeded in defragmenting all given endpoints.
```

File: etcdutl/README.md [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+
+Exit status '0' when the process was successful.
+
+#### Example
+
+To defragment a data directory directly, use the `--data-dir` flag:
+
+``` bash
+# Defragment while etcd is not running
+./etcdctl defrag --data-dir default.etcd
```
Comment (lilic) [inline etcdutl/README.md:30]: Did you mean etcdutl here?

```suggestion
./etcdutl defrag --data-dir default.etcd
```

File: etcdutl/etcdutl/backup_command.go [hunk 5]
```diff
@@ -0,0 +1,331 @@
+// Copyright 2015 The etcd Authors
```
Comment (lilic) [inline etcdutl/etcdutl/backup_command.go:1]: Should we change this as well, as while some things were copied a lot of the code changed?

File: CHANGELOG-3.5.md [hunk 6]
```diff
@@ -223,6 +223,8 @@ Note that any `etcd_debugging_*` metrics are experimental and subject to change.
 - Add [`etcdctl get --count-only`](https://github.com/etcd-io/etcd/pull/11743) flag for output type `fields`.
 - Add [`etcdctl member list -w=json --hex`](https://github.com/etcd-io/etcd/pull/11812) flag to print memberListResponse in hex format json.
 - Changed [`etcdctl lock <lockname> exec-command`](https://github.com/etcd-io/etcd/pull/12829) to return exit code of exec-command.
+- [New tool: `etcdutl`](https://github.com/etcd-io/etcd/pull/12971) incorporated functionality of: `etcdctl snapshot status|restore`, `etcdctl backup`, `etcdctl`.-
```
Comment (lilic) [inline CHANGELOG-3.5.md:226]: Should we mention defrag as well?

**Review by lilic — COMMENTED** [review]
Wow amazing work here! I briefly had a look.

Do you think there is any risk in including this in 3.5? From what I can tell, a lot of things were mainly moved over and nicely cleaned up, so risk is low in bugs?

**Comment by ptabor** [comment]
@hexfusion I hope that the change is complete, in particular it contains: 
  - changes to release script to take in consideration etcdutl binary existence (tested) 
  - e2e tests executed against both: etcdctl & etcdutl variants of the commands
  - README.md

File: etcdutl/etcdutl/backup_command.go (hunk 5 above)
Comment (ptabor) [reply etcdutl/etcdutl/backup_command.go]: DONE

File: CHANGELOG-3.5.md (hunk 6 above)
Comment (ptabor) [reply CHANGELOG-3.5.md]: Good catch. Done.

**Comment by ptabor** [comment]
> Wow amazing work here! I briefly had a look.
> 
> Do you think there is any risk in including this in 3.5? From what I can tell, a lot of things were mainly moved over and nicely cleaned up, so risk is low in bugs?

Thank you for review. I don't expect any changes in the pre-existing functionality of etdctl (except v2 migrate that is gone). The functionality is covered with e2e tests.   So I don't think its a risky change.

**Comment by lilic** [comment]
Overall from a code perspective and dev perspective, this makes perfect sense. 👍 

My only concern is that this would be making the user experience a bit more complex, as users would need 3 binaries/tools to run and manage etcd. But we can see how it goes and gather user feedback when end-users start using `etcdutl`?

**Review by hexfusion — APPROVED** [review]
> The etcdctl commands keep being served as they used to, but just print additional warning with recommendation to migrate to etcdutl.

With the above messaging, I feel this change is reasonable. By separating these very admin-specific actions of backing up and restoring state to a separate binary we provider at least basic isolation. The alternative would be RBAC level controls similar to `oc adm` but since etcd RBAC will probably never be a part of kubernetes 1.x +1.

**Comment by ptabor** [comment]
Thank you for quick review.

**Comment by gyuho** [comment]
Highlighted this in the 3.5 blog as well.

**Pull request merged**